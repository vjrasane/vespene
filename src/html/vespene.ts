

import { Maybe } from "./maybe";


namespace Vespene {
	export type Hooks = {};

	export type FunctionComponent<
		P extends object = {},
		> = (props: WithChildren<P>, hooks: Hooks) => Element | null;

	export type Node = Element | string | null | undefined;

	export type HTMLElementTag = keyof HTMLElementTagNameMap;

	type DeepArray<T> = (T | DeepArray<T>)[];

	export type Children<C extends Node = Node> = C | DeepArray<C> | undefined

	export type WithChildren<P extends object, C extends Node = Node> = {
		children?: Children<C>
	} & P;

	const flatten = <T>(arr: DeepArray<T>): Array<T> => {
		return arr.flatMap(
			(elem) => Array.isArray(elem) ? flatten(elem) : elem
		)
	}

	const compact = <T>(arr: Array<Maybe<T>>): Array<T> => {
		return arr.filter((elem): elem is T => elem != null);
	}

	export abstract class Element<
		C extends Node = Node
		> {
		protected children: Array<Node>;
		protected parent: Maybe<Element>;

		protected cleanupHandlers: Array<() => void> = [];

		constructor(children: Array<C>) {
			this.children = children;
		}

		abstract render: (parent: Maybe<Element>, force?: boolean) => Maybe<HTMLElement>;
		abstract getElement: () => Maybe<HTMLElement>;
		protected abstract clearElement: () => void;

		protected removeChild = (node: Node): void => {
			const index = this.children.findIndex((child) => child === node);
			if (index < 0) return;
			this.children.splice(index, 1);
		};

		protected replaceChild = (node: Node, replacer: Node): void => {
			const index = this.children.findIndex((child) => child === node);
			if (index < 0) return;
			this.children.splice(index, 1, replacer);
		};

		protected addChild = (node: Node): void => {
			const index = this.children.findIndex((child) => child === node);
			if (index >= 0) return;
			this.children.push(node);
		}

		protected runCleanup = (): void => {
			this.cleanupHandlers.forEach((handler) => handler());
			this.children.forEach(
				(child) => child && typeof child !== "string" && child.runCleanup()
			)
			this.clearElement();
		}

		cleanup = (handler: () => void): void => {
			this.cleanupHandlers.push(handler);
		};

		remove = (): void => {
			this.getElement()?.remove();
			this.parent?.removeChild(this);
			this.runCleanup();
		};

		replace = <N extends Node>(node: N): N => {
			const rendered = typeof node === "string"
				? node
				: node?.render(this, false);
			if (!rendered) {
				this.remove();
			} else {
				this.getElement()?.replaceWith(rendered);
				this.parent?.replaceChild(this, node);
				this.runCleanup();
			}
			return node;
		};

		append = <N extends Node>(node: N): N => {
			this.children.push(node);
			const rendered = typeof node === "string"
				? node
				: node?.render(this, false);
			if (!rendered) return node;
			this.getElement()?.append(rendered);
			return node;
		}

		appendTo = (parent: Element): Element => {
			this.parent?.removeChild(this);
			this.parent = parent;
			this.parent.addChild(this);
			const rendered = this.render(parent, false);
			if (!rendered) return this;
			this.parent?.getElement()?.append(rendered);
			return this;
		}
	}

	class ActualElement<
		C extends Node = Node,
		P extends object = {}> extends Element<C> {
		private tag: string;
		private props: P;

		private element: Maybe<HTMLElement>;

		constructor(tag: string, props: P, children: Array<C>) {
			super(children);
			this.tag = tag;
			this.props = props;
		}

		getElement = (): Maybe<HTMLElement> => {
			return this.element;
		};

		protected clearElement = (): void => {
			this.element = undefined;
		}

		render = (parent: Maybe<Element>, force = true): Maybe<HTMLElement> => {
			this.parent = parent;
			if (this.element && !force) return this.element;
			this.element = document.createElement(this.tag);
			const children = compact(this.children.map(
				child => typeof child === "string"
					? child
					: child?.render(this, force)
			));
			this.element?.append(...children);
			return this.element;
		}
	}

	class FunctionElement<
		C extends Node = Node,
		P extends object = {}
		> extends Element<C> {

		private func: FunctionComponent<P>;
		private props: P;

		private element: Maybe<Element>;

		constructor(func: FunctionComponent<P>, props: P, children: Array<C>) {
			super(children)
			this.func = func;
			this.props = props;
		}

		protected clearElement = (): void => {
			this.element = undefined;
		}

		getElement = (): Maybe<HTMLElement> => {
			return this.element?.getElement();
		};

		render = (parent: Maybe<Element>, force = true): Maybe<HTMLElement> => {
			this.parent = parent;
			if (this.element && !force) return this.getElement();
			this.element = this.func(
				{ ...this.props, children: this.children }, {});
			return this.element?.render(this, force);
		};
	}
	export const createElement = <P extends object, C extends Node>(
		component: FunctionComponent<P> | string,
		props: P,
		...children: DeepArray<C>
	): Element => {
		return typeof component === "string"
			? new ActualElement(component, props, flatten(children))
			: new FunctionElement(component, props, flatten(children))
	}

	export const render = (root: HTMLElement, element: Vespene.Element) => {
		const rendered = element.render(undefined);
		if (!rendered) return;
		root.append(rendered);
	};
}

export default Vespene