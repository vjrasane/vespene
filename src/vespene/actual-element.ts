import Element from "./element";
import { Maybe } from "./types";
import { compact, isEventHandler } from "./utils";

export default class ActualElement<
	P extends object = {}> extends Element<P, string, HTMLElement> {

	private bindAttributes = (): void => {
		Object.entries(this.props ?? {}).filter(
			([key, value]) => !isEventHandler(key, value)
		).forEach(
			([key, value]) => {
				this.element?.setAttribute(key, value);
			}
		);
	}

	private bindEventListeners = (): void => {
		Object.entries(this.props ?? {}).filter(
			([key, handler]) => isEventHandler(key, handler)
		).forEach(
			([key, handler]) => {
				const event = key.substring(2).toLowerCase();
				this.element?.addEventListener(event, handler);
			}
		);
	}

	getElement = (): Maybe<HTMLElement> => {
		return this.element;
	};

	render = (force = true): Maybe<HTMLElement> => {
		if (this.element && !force) return this.element;
		this.element = document.createElement(this.type);
		const children = compact(this.children.map(
			child => child instanceof Element
				? child?.render(force)
				: child
		));
		this.element?.append(...(children as Array<string | HTMLElement>));
		this.bindAttributes();
		this.bindEventListeners();
		return this.element;
	}
}