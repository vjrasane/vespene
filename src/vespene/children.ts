
import { Node } from "./types";
import { flatten, compact } from "./utils";

export default {
	toArray: (children: Node | Node[]): Array<Exclude<Node, boolean | null | undefined>> => {
		return compact(flatten([children]));
	}
}