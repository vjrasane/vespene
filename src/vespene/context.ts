import { createElement, Fragment, Element, FunctionComponent } from ".";
import { PropsWithChildren } from "./types";


type ProviderProps<T> = {
	value: T;
  };

type ConsumerProps<T> = {
	children: (value: T) => Element
}

const createContextProvider = function<T>(context: Context<T>): FunctionComponent<ProviderProps<T>> {
	return ({ value, children }, { setContext }) => {
		setContext(context, value);
		return createElement(Fragment, null, children);
	}
}

const createContextConsumer = function<T>(context: Context<T>): FunctionComponent<ConsumerProps<T>> {
	return ({ children }, { getContext }) => children(getContext(context))
}

export class Context<T> {
	readonly id: symbol = Symbol();
	readonly defaultValue: T;

	readonly Provider: FunctionComponent<ProviderProps<T>>;
	readonly Consumer: FunctionComponent<ConsumerProps<T>>;

	constructor(defaultValue: T) {
		this.defaultValue = defaultValue;
		this.Provider = createContextProvider(this);
		this.Consumer = createContextConsumer(this);
	}
}
