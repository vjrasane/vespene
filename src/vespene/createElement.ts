import Element from "./element"
import { Fragment, Node, Child, FunctionComponent, Maybe } from "./types"
import { flatten } from "./utils";

export const createElement = <P extends object>(
	component: string | typeof Fragment | FunctionComponent<P>,
	props: Maybe<P | null>,
	...children: Array<Node>
): Element => {
	return new Element(component, props ?? {} as P, 
		children.length > 1 ? children : children[0]);
}