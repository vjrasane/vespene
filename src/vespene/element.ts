import { Context } from "./context";
import { Fragment, Node, Child, Text, Maybe, PropsWithChildren, FunctionComponent } from "./types";
import { compact, flatten, isEventHandler, removeBy, toCssStyle } from "./utils";

// const renderSuccessors = (successors: HTMLElement | Array<Child>, parent: Element): Array<HTMLElement | Text> => {
// 	return Array.isArray(successors) 
// 	? successors.flatMap(
// 		successor => successor instanceof Element ? successor.render(parent) : successor) 
// 	: [successors];
// }

export default class Element<P = any, 
	T extends string | typeof Fragment | FunctionComponent<P> = string | typeof Fragment | FunctionComponent<P>> {
	private readonly instructions: {
		readonly type: T;
		readonly props: P;
		readonly children: Node;
	};

	// protected elements: Maybe<Array<Element | Text | HTMLElement>>;
	protected parent: Maybe<Element | HTMLElement>;

	protected cleanupHandlers: Array<() => void> = [];

	// protected element: Maybe<Element | Text | HTMLElement | null>;

	// protected successors: Maybe<HTMLElement | Element | Array<Child> | null>;

	protected context: Record<symbol, any> = {};

	protected element: Maybe<HTMLElement | null>;

	protected children: Maybe<Array<Child>>;

	constructor(
		type: T,
		props: P,
		children: Node) {
			this.instructions = {
				type, 
				props,
				children
			};
	}

	get type() {
		return this.instructions.type;
	}

	get props(): Readonly<P> {
		return this.instructions.props;
	}

	// private getContainer = (parent: Maybe<Element | HTMLElement>): Maybe<HTMLElement> => {
	// 	if (!parent) return;
	// 	if (parent instanceof HTMLElement) return parent;
	// 	if (parent.element instanceof HTMLElement) return parent.element;
	// 	return this.getContainer(parent.parent);
	// }

	private isParent = (child: Element): boolean => {
		if (child === this) return true;
		if (!child.parent) return false;
		if (!(child.parent instanceof Element)) return false;
		return this.isParent(child.parent);
	}

	private bindAttribute = (key: string, value: unknown) => {
		switch (key) {
			case "style":
				this.setStyle(value as Record<string, string | number>);
				break;
			default:
				this.setAttribute(key, value as string | number);
				break;
		}
	}

	private bindAttributes = (): void => {
		Object.entries(this.props).filter(
			([key, value]) => !isEventHandler(key, value)
		).forEach(([key, value]) => this.bindAttribute(key, value));
	}

	private bindEventListeners = (element: HTMLElement): void => {
		Object.entries(this.props).filter(
			([key, handler]) => isEventHandler(key, handler)
		).forEach(
			([key, handler]) => {
				const event = key.substring(2).toLowerCase();
				element.addEventListener(event, handler as EventListener);
			}
		);
	}

	private renderChildren = (children: Array<Child>): Array<HTMLElement | string> => {
		children.forEach((child) => { 
			if(!(child instanceof Element )) return; 
			child.parent = this;
		 });
		return compact(children.flatMap(
			(child): Array<HTMLElement | Text | null> => child instanceof Element 
				? child?.render(this.context) : [child]
		)).map(child => child instanceof HTMLElement ? child : String(child));
	}

	private renderActual = (tag: string): HTMLElement => {
		this.element = document.createElement(tag);
		this.children = flatten([this.instructions.children]) as Array<Child>;
		this.element.append(...this.renderChildren(this.children));
		this.bindAttributes();
		this.bindEventListeners(this.element);
		return this.element;
	}

	private renderFunction = (func: FunctionComponent<P>): Array<HTMLElement | string> => {
		const props: PropsWithChildren<P> = { ...this.props, children: this.instructions.children };
		const elements = func(props, this);
		this.children = Array.isArray(elements) ? elements : compact([elements]);
		return this.renderChildren(this.children);
	}

	protected getSuccessor = (): Maybe<HTMLElement | Element> => {
		if (this.element instanceof HTMLElement) return this.element;
		if (this.element === null) return;
		if (!this.children) return;
		if (this.children.length !== 1) return;
		const [element] = this.children;
		if (element instanceof Element) return element;
	}

	protected getElement = (): Maybe<HTMLElement> => {
		const successor = this.getSuccessor();
		if (successor instanceof HTMLElement) return successor;
		if (successor instanceof Element) return successor.getElement();
	}

	protected removeChild = (node: Element): void => {
		if (!this.children) return;
		this.children = removeBy(
			(child) => child instanceof Element && child === node,
			this.children
		);
	};

	protected replaceChild = (node: Element, replacer: Child): void => {
		if (!this.children) return;
		const index = this.children.findIndex(
			(child) => child instanceof Element && child === node
		);
		if (index < 0) return;
		this.children.splice(index, 1, replacer);
	};

	protected addChild = (node: Child): void => {
		if (!this.children) return;
		const index = this.children.findIndex(
			(child) => child instanceof Element && child === node);
		if (index >= 0) return;
		this.children.push(node);
	}

	private runCleanup = (): void => {
		this.cleanupHandlers.forEach((handler) => handler());
		this.cleanupHandlers = [];
		this.element = null;
		this.children?.forEach(
			child => child instanceof Element && child.runCleanup());
		this.children = [];
	}

	setStyle = (style: Record<string, string | number>) => {
		this.setAttribute("style", toCssStyle(style))
	}

	setAttribute = (key: string, value: string | number) => {
		this.getElement()?.setAttribute(key, `${value}`);
	}

	cleanup = (handler: () => void): void => {
		this.cleanupHandlers.push(handler);
	};

	remove = (): void => {
		const element = this.getElement();
		if (!element) return;
		if (this.parent instanceof Element) this.parent.removeChild(this);
		this.parent = undefined;
		this.runCleanup();
	};

	private replaceWithElement = (replacer: Element): void => {
		this.element = replacer.element;
		this.children = replacer.children;
		this.cleanupHandlers = replacer.cleanupHandlers;
	}

	replace = (node: Element): void => {
		const element = this.getElement();
		if (!element) return;
		const replaceable = node.render(this.context);
		if (!replaceable) return this.remove();
		element.replaceWith(...replaceable);
		this.runCleanup();
		this.replaceWithElement(node);
	};

	protected moveToParent = (parent: Element): void => {
		if (parent === this) return;
		if (this.parent instanceof Element) this.parent?.removeChild(this);
		parent.addChild(this);
		this.parent = parent;
	}

	append = (node: Child): void => {
		const successor = this.getSuccessor();
		if (!successor) return;
		if (successor instanceof Element) return successor.append(node);
		const appendable = node instanceof Element ? node.render(this.context) : String(node);
		if (appendable)	successor.append(...appendable);
		if (node instanceof Element) node.moveToParent(this);
	}

	appendTo = (parent: Element | HTMLElement): void => {
		if (parent instanceof Element) return parent.append(this);
		const appendable = this.render(this.context);
		if (appendable) parent.append(...appendable);
		this.parent = parent;
	}

	setContext = <T>(context: Context<T>, value: T): void => {
		this.context = { ...this.context, [context.id]: value };
	}

	getContext = <T>(context: Context<T>): T => {
		if (!this.context || !(context.id in this.context)) return context.defaultValue;
		return this.context[context.id];
	}

	private render = (context: Record<symbol, any>): Array<HTMLElement | string> => {
		switch (this.children) {
			case undefined: {
				this.context = { ...context };
				switch (typeof this.type) {
					case "string":
						return [this.renderActual(this.type)];
					case "function":
						return this.renderFunction(this.type);
					case "symbol":
						this.children = flatten([this.instructions.children]) as Array<Child>;
						return this.renderChildren(this.children);
				}
			}
			default:
				switch (this.element) {
					case undefined:
						return this.renderChildren(this.children);
					case null:
						return [];
					default:
						return [this.element];
				}
		}	
	};
}
