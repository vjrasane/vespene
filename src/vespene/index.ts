


import Element from "./element";
import { FunctionComponent, Fragment } from "./types";
import { createElement } from "./createElement";
import { render } from "./render";
import Children from "./children";

export { createElement, render, Children, Element, Fragment, type FunctionComponent };

export type HTMLElementTag = keyof HTMLElementTagNameMap;




