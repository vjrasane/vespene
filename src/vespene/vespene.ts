


import Element from "./element";
import { FunctionComponent } from "./types";
import { createElement } from "./createElement";
import { render } from "./render";
import Children from "./children";
export { createElement, render, Children, Element, type FunctionComponent };

export type HTMLElementTag = keyof HTMLElementTagNameMap;




