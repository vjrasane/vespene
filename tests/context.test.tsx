import V, { FunctionComponent, render } from "../src";
import { Context } from "../src/vespene/context";

describe("context", () => {
	it("uses default value if context not set", () => {
		const Ctx = new Context<string | undefined>("default");
		render(document.body, 
				<Ctx.Consumer>
					{(value) => <div>{value}</div>}
				</Ctx.Consumer>
		);
		expect(document.body).toMatchSnapshot();
	})

	it("uses default value in function component if context not set", () => {
		const Ctx = new Context<string | undefined>("default");
		const Component : FunctionComponent = (props, { getContext }) => <div>{getContext(Ctx)}</div>;
		render(document.body, 
				<Component />
		);
		expect(document.body).toMatchSnapshot();
	})

	it("passes correct value to child", () => {
		const Ctx = new Context<string | undefined>(undefined);
		render(document.body, 
		<Ctx.Provider value="context">
			<Ctx.Consumer>
				{(value) => <div>{value}</div>}
			</Ctx.Consumer>
		</Ctx.Provider>);
		expect(document.body).toMatchSnapshot();
	});

	it("passes innermost provider value to child", () => {
		const Ctx = new Context<string | undefined>(undefined);
		render(document.body, 
		<Ctx.Provider value="first">
			<Ctx.Provider value="second">
				<Ctx.Provider value="third">
					<Ctx.Consumer>
						{(value) => <div>{value}</div>}
					</Ctx.Consumer>
				</Ctx.Provider>
			</Ctx.Provider>
		</Ctx.Provider>
		);
		expect(document.body).toMatchSnapshot();
	});

	it("passes correct value after appending", () => {
		const Ctx = new Context<string | undefined>(undefined);
		const element = <div>Element</div>;
		render(document.body, 
		<Ctx.Provider value="context">
			{element}
		</Ctx.Provider>);
		element.append(	<Ctx.Consumer>
			{(value) => <div>{value}</div>}
		</Ctx.Consumer>)
		expect(document.body).toMatchSnapshot();
	});

	it("passes correct value after appending to", () => {
		const Ctx = new Context<string | undefined>(undefined);
		const element = <div>Element</div>;
		render(document.body, 
		<Ctx.Provider value="context">
			{element}
		</Ctx.Provider>);
		(<Ctx.Consumer>
			{(value) => <div>{value}</div>}
		</Ctx.Consumer>).appendTo(element)
		expect(document.body).toMatchSnapshot();
	})

	it("passes correct value after replacing", () => {
		const Ctx = new Context<string | undefined>(undefined);
		const element = <div>Element</div>;
		render(document.body, 
		<Ctx.Provider value="context">
			{element}
		</Ctx.Provider>);
		element.replace(<Ctx.Consumer>
			{(value) => <div>{value}</div>}
		</Ctx.Consumer>);
		expect(document.body).toMatchSnapshot();
	});

	it("passes correct value to function component", () => {
		const Ctx = new Context<string | undefined>(undefined);
		const Component : FunctionComponent = (props, { getContext }) => <div>{getContext(Ctx)}</div>;
		render(document.body, 
			<Ctx.Provider value="context">
				<Component/>
			</Ctx.Provider>);
			expect(document.body).toMatchSnapshot();
	});
})