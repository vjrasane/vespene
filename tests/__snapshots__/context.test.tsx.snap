// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context passes correct value after appending 1`] = `
<body>
  <div>
    Element
    <div>
      context
    </div>
  </div>
</body>
`;

exports[`context passes correct value after appending to 1`] = `
<body>
  <div>
    Element
    <div>
      context
    </div>
  </div>
</body>
`;

exports[`context passes correct value after replacing 1`] = `
<body>
  <div>
    context
  </div>
</body>
`;

exports[`context passes correct value to child 1`] = `
<body>
  <div>
    context
  </div>
</body>
`;

exports[`context passes correct value to function component 1`] = `
<body>
  <div>
    context
  </div>
</body>
`;

exports[`context passes innermost provider value to child 1`] = `
<body>
  <div>
    third
  </div>
</body>
`;

exports[`context uses default value if context not set 1`] = `
<body>
  <div>
    default
  </div>
</body>
`;

exports[`context uses default value in function component if context not set 1`] = `
<body>
  <div>
    default
  </div>
</body>
`;
